### PR TITLE
MINOR: Improve IntegrationTestUtils documentation

### PR DESCRIPTION
**Pull request by dongjinleekr**
While I was writing [an example project on testing Kafka Streams](https://github.com/dongjinleekr/kafka-streams-example), I found that there are several problems with the documentation of `IntegrationTestUtils.`

- The documentation of `Time` parameter in `produceKeyValuesSynchronously` methods is missing. This parameter was added in commit de1b853, but documentation was omitted then.
- Change parameter `enableTransactions` in `produceKeyValuesSynchronouslyWithTimestamp(String, Collection, Properties, Headers headers, Long, boolean)` into `enableTransactions`: for consistency with the other overloads.
- Add Javadoc to undocumented methods, like `produceKeyValuesSynchronouslyWithTimestamp`, `waitUntilMinRecordsReceived`, etc.
- Fix ordering: `waitUntilMinRecordsReceived`
- Upper bound in ... -> Upper bound of ...

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by vvcephei — APPROVED** [review]
Thanks for the cleanup @dongjinleekr !

This looks good to me.

**Comment by dongjinleekr** [comment]
Is there anyone who can merge this PR into 2.1.0 release?

**Comment by dongjinleekr** [comment]
@lindong28 Excuse me. Could this be included in 2.1.0?

**Comment by dongjinleekr** [comment]
Cc/ @cmccabe @ijuma @guozhangwang

**Comment by dongjinleekr** [comment]
@vahidhashemian Excuse me. Could have a look?

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java [hunk 1]
```diff
@@ -195,6 +224,33 @@ public static void cleanStateAfterTest(final EmbeddedKafkaCluster cluster, final
...
+     * @param headers             {@link Headers} of the data records
+     * @param timestamp           Timestamp of the record
+     * @param <K>                 Key type of the data records
+     * @param <V>                 Value type of the data records
+     */
+    public static <K, V> void produceKeyValuesSynchronouslyWithTimestamp(final String topic,
+                                                                         final Collection<KeyValue<K, V>> records,
+                                                                         final Properties producerConfig,
+                                                                         final Headers headers,
+                                                                         final Long timestamp)
```
Comment (vahidhashemian) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: Was there a specific reason for adding this method?

**Review by vahidhashemian — COMMENTED** [review]
Thanks for the PR. Just one question inline.

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java (hunk 1 above)
Comment (dongjinleekr) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: For parity. `produceKeyValuesSynchronouslyWithTimestamp` has 3 required and 3 optional parameters[^1]. By adding this method (overload 2), we can provide a counterpart of overload 4. (see below) Additionally, I found that this method is handier than using overload 4 for header parameters.

|          | headers | timestamp | enableTransactions |
|----------|---------|-----------|--------------------|
|     1    |    X    |     O     |          X         |
| 2 (new!) |    O    |     O     |          X         |
|     3    |    X    |     O     |          O         |
|     4    |    O    |     O     |          O         |

[^1]: topic, records, producerConfig / headers, timestamp, enableTransactions.

Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: That makes sense, but is this method currently unused?

If it's not used, then I think it's better not to add it. (IMHO, lack of dead code outweighs the value of symmetry)

**Review by vvcephei — COMMENTED** [review]
Hi @dongjinleekr ,

Sorry for the silence... I think the only controversial portion is the new test util method. Otherwise, this PR has my approval (for what it's worth).

Maybe @guozhangwang can take a look?

-John

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java (hunk 1 above)
Comment (dongjinleekr) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: @vvcephei Thank you for your comment. Yes, you are right - this method is not used in the project. But `IntegrationTestUtils` is also intended to be provided to the end users [as a testing utility for Kafka Streams applications, with `EmbeddedKafkaCluster`.](https://github.com/dongjinleekr/kafka-streams-example/blob/master/src/test/java/com/dongjinlee/examples/kafkastreams/IntegrationTest.java)

For this reason, I thought adding this method is reasonable, with documentation for the end users.

Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: Thanks for the response, @dongjinleekr .

I'm not too sure if `IntegrationTestUtils` is _intended_ to be a public API, or if it's just been used that way, since there was no good way to write integration tests without `EmbeddedKafkaCluster`. I think that example is more of a symptom of how strong the need is, with no good solution available.

I've heard people mention in the past that there's some desire to offer a real public integration test api, but I couldn't find any evidence of it on Jira. I've proposed https://issues.apache.org/jira/browse/KAFKA-7852 , so maybe we can do something about it.

Regardless, generally speaking, a project's own tests shouldn't be considered a public API. So, while I see your point about the de facto state of things, it still doesn't seem quite right to start adding methods specifically for public use.

Plus there a catch-22: if it's _not_ a public API, then we can make any change we want in a PR, but if it _is_ a public API, then we must go through a KIP to propose this change ;)

Comment (vahidhashemian) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: Even though I have no strong opinions here, I agree more with @vvcephei. If we have an unused function chances are though some cleanup PR in the future it'll get removed. I personally prefer not to add unused functions. I hope that makes sense.

Comment (dongjinleekr) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: As far as I know, `IntegrationTestUtils` and `EmbeddedKafkaCluster` are intended to be used as a testing utility. Here is an excerpt from [Kafka Streams in Action](https://www.manning.com/books/kafka-streams-in-action) by @bbejeck. As you can see, both classes are used for integration tests for Kafka Streams application.

streams-test

However, I totally agree that there is no good way for testing Kafka Streams applications, in an end user's perspective. I experienced several questions about how to write testing code from the Kafka communities in Seoul and Tokyo. (In fact, the example above was originally written to show them.) @vvcephei, You are right - this problem reflects the strong need for testing utility, not only violates the principle that "a project's own tests shouldn't be considered a public API" or not have to be exposed for public use.

It seems like there was a plan to develop the testing utility module, but what I could find was [a draft KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-139%3A+Kafka+TestKit+library) only. I think this topic should be discussed after 2.2.0 release.

So, in short, how about this?

1. Remove this new overload (i.e., keep the status quo.)
2. Merge [KAFKA-7852](https://issues.apache.org/jira/browse/KAFKA-7852) into [KAFKA-5041](https://issues.apache.org/jira/browse/KAFKA-5041) by @ijuma (As @mjsax pointed out, it is a duplication.)
3. Continue to discuss this topic with the KIP later.

Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: Oops! I've closed that ticket. I didn't find KAFKA-5041 when I searched for it.

I think your plan is a good one.

I certainly won't dispute the fact that we have widely (but pathologically) recommended people to use the test artifact for their own tests. Still, I agree that it would be more productive to push for KAFKA-5041/KIP-139, rather than to start building explicitly public APIs into the (should be internal) testing utilities.

Thanks,
-John

**Comment by dongjinleekr** [comment]
@vvcephei @vahidhashemian Here it is. I removed the method.

**Comment by mjsax** [comment]
@dongjinleekr Kafka 2.1.0 was release 4 month ago. We can get it into 2.2.0 thought. Also, it's only test code, so we could, if we want back port it to get it into 2.1.1 -- however, @cmccabe is cutting 2.1.1-rc0 soon, so it might not make it -- also not worth blocking 2.1.1 release for this.

Retest this please.

**Comment by cmccabe** [comment]
@mjsax: agreed.

**Comment by dongjinleekr** [comment]
@mjsax @cmccabe No problem. Thanks for your comment!

**Review by vahidhashemian — APPROVED** [review]
LGTM. Thanks for addressing the earlier comments.
